src/uct/tcp/tcp_ep.c

```diff
@@ -877,6 +1003,81 @@ uct_tcp_ep_am_sendv(uct_tcp_iface_t *iface, uct_tcp_ep_t *ep,
     return status;
 }
 
+static inline void uct_tcp_ep_send_put_ack(uct_tcp_iface_t *iface,
+                                           uct_tcp_ep_t *ep,
+                                           uct_tcp_am_hdr_t *hdr)
+{
+    ucs_assertv(hdr != NULL, "ep=%p", ep);
+
+    hdr->am_id  = UCT_TCP_EP_PUT_ACK_AM_ID;
+    hdr->length = 0;
+    uct_tcp_ep_am_send(iface, ep, hdr);
+}
+
+static ucs_status_t
+uct_tcp_ep_pending_put_ack_cb(uct_pending_req_t *self)
+{
+    uct_tcp_am_hdr_t *hdr = NULL;
+    uct_tcp_ep_put_ack_pending_req_t *put_ack_req;
+    uct_tcp_iface_t *iface;
+    ucs_status_t status;
+
+    put_ack_req = ucs_derived_of(self, uct_tcp_ep_put_ack_pending_req_t);
+    iface       = ucs_derived_of(put_ack_req->ep->super.super.iface,
+                                 uct_tcp_iface_t);
+
+    status = uct_tcp_ep_am_prepare(iface, put_ack_req->ep, 0, &hdr);
+    if (ucs_likely(status == UCS_OK)) {
+        uct_tcp_ep_send_put_ack(iface, put_ack_req->ep, hdr);
+        ucs_free(put_ack_req);
+        return UCS_OK;
+    } else if (status == UCS_ERR_NO_RESOURCE) {
+        return UCS_INPROGRESS;
+    }
+
+    return status;
+}
+
+static void uct_tcp_ep_post_put_ack(uct_tcp_ep_t *ep,
+                                    uct_tcp_ep_put_req_t *put_req)
+{
+    uct_tcp_am_hdr_t *hdr  = NULL;
+    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                            uct_tcp_iface_t);
+    uct_tcp_ep_put_ack_pending_req_t *put_ack_req;
+    ucs_status_t status;
+
+    /* Make sure that we are sending nothing through this EP at the moment.
+     * This check is needed to avoid mixing AM/PUT data sent from this EP
+     * and this PUT ACK message */
+    status = uct_tcp_ep_am_prepare(iface, ep, 0, &hdr);
+    if (ucs_likely(status == UCS_OK)) {
+        uct_tcp_ep_send_put_ack(iface, ep, hdr);
```

Comment:
add `return` here and last `else` will not be needed

Reply:
done